File: neetcode-250/trees/110-balanced-binary-tree.ts
import { TreeNode } from "./treeNode";

interface NodeRes {
	balanced: boolean;
	height: number;
}

function isBalanced(root: TreeNode | null): boolean {
	return dfs(root).balanced;

	function dfs(node: TreeNode | null): NodeRes {
		if (!node) {
			return { balanced: true, height: 0 };
		}

		const left = dfs(node.left);
		const right = dfs(node.right);
		const balanced = left.balanced && right.balanced && Math.abs(left.height - right.height) <= 1;
		return { balanced, height: 1 + Math.max(left.height, right.height) };
	}

	function getHeight(res: (boolean | number)[]): number {
		return res[1] as number;
	}
}

File: neetcode-250/trees/110-balanced-binary-tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return (-1, True)

            left_hei, left_bal = dfs(node.left)
            if not left_bal:
                return 0, False
            right_hei, right_bal = dfs(node.right)
            return (
                1 + max(left_hei, right_hei),
                left_bal and right_bal and abs(left_hei - right_hei) <= 1,
            )

        return dfs(root)[1]
